airbyte-integrations/connectors/source-harvest/source_harvest/streams.py

```diff
@@ -0,0 +1,299 @@
+        for item in items.read_records(sync_mode=sync_mode):
+            yield {"parent_id": item["id"]}
+
+    def path(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> str:
+        return self.path_template.format(parent_id=stream_slice["parent_id"])
+
+
+class Contacts(IncrementalHarvestStream):
+    pass
+
+
+class Clients(IncrementalHarvestStream):
+    pass
+
+
+class Company(HarvestStream):
+    primary_key = None
+    data_field = None
+
+
+class Invoices(IncrementalHarvestStream):
+    pass
+
+
+class InvoiceMessages(HarvestSubStream, IncrementalHarvestStream):
+    parent_stream = Invoices
+    path_template = "invoices/{parent_id}/messages"
+
+
+class InvoicePayments(HarvestSubStream, IncrementalHarvestStream):
+    parent_stream = Invoices
+    path_template = "invoices/{parent_id}/payments"
+
+
+class InvoiceItemCategories(IncrementalHarvestStream):
+    pass
+
+
+class Estimates(IncrementalHarvestStream):
+    pass
+
+
+class EstimateMessages(HarvestSubStream, IncrementalHarvestStream):
+    parent_stream = Estimates
+    path_template = "estimates/{parent_id}/messages"
+
+
+class EstimateItemCategories(IncrementalHarvestStream):
+    pass
+
+
+class Expenses(IncrementalHarvestStream):
+    pass
+
+
+class ExpenseCategories(IncrementalHarvestStream):
+    pass
+
+
+class Tasks(IncrementalHarvestStream):
+    pass
+
+
+class TimeEntries(IncrementalHarvestStream):
+    pass
+
+
+class UserAssignments(IncrementalHarvestStream):
+    pass
+
+
+class TaskAssignments(IncrementalHarvestStream):
+    pass
+
+
+class Projects(IncrementalHarvestStream):
+    pass
+
+
+class Roles(IncrementalHarvestStream):
+    pass
+
+
+class Users(IncrementalHarvestStream):
+    pass
+
+
+class BillableRates(HarvestSubStream):
+    parent_stream = Users
+    path_template = "users/{parent_id}/billable_rates"
+
+
+class CostRates(HarvestSubStream):
+    parent_stream = Users
+    path_template = "users/{parent_id}/cost_rates"
+
+
+class ProjectAssignments(HarvestSubStream, IncrementalHarvestStream):
+    parent_stream = Users
+    path_template = "users/{parent_id}/project_assignments"
+
+
+class ReportsBase(HarvestStream, ABC):
+    data_field = "results"
+    date_param_template = "%Y%m%d"
+
+    @property
+    @abstractmethod
+    def report_path(self):
+        """
+        :return: report path suffix
+        """
+
+    def __init__(self, from_date: pendulum.date = None, **kwargs):
+        super().__init__(**kwargs)
+        self._from_date = from_date or pendulum.now().date().subtract(years=1)
+
+    def request_params(self, **kwargs) -> MutableMapping[str, Any]:
+        params = super().request_params(**kwargs)
+        current_date = pendulum.now()
+        # `from` and `to` params are required for reports calls
+        # min `from` value is current_date - 1 year
+        params.update({"from": self._from_date.strftime("%Y%m%d"), "to": current_date.strftime("%Y%m%d")})
```

Comment:
any reason not to make this incremental by manipulating the `from` and `to` params?

Reply:
Yes, we have a reason for that. Discussed [there](https://github.com/airbytehq/airbyte/pull/3709#discussion_r642612854).

Reply:
we can enrich the record returned from the API with this info by doing the following: 

1. inside `parse_response` we can access the original `request` object that was made then parse the `from` and `to` parameters. 
2. put those properties in the record response i.e: `record['from'] = ...`
3. parse inside the `get_updated_state` method
4. profit
WDYT?

Reply:
this would require changing the schema too

Reply:
It makes sense for me. I'll try to do it.